src/wwwroot/app/directives/candidate-vacancy-info/candidate-vacancy-info.js

```diff
@@ -213,93 +269,255 @@ function CandidateVacancyInfoController($scope, $translate, $q, $state, UserDial
+         stage: rejectStage.stage,
+         stageId: rejectStage.stage.id,
+         //TODO: open dialog to write a comment
+         comment: { message: 'rejected' },
+         stageState: STAGE_STATES.Active,
+         createdOn: (new Date()).toISOString()
+      };
+      candidateStage.vacancyStageInfos.push(rejectedVsi);
+      recalculateCurrentStageId(candidateStage);
+      vm.stagesToShow = calculateVacancyStagesEntitiesCount(vm.vacancyStageInfosComposedByCandidateIdVacancyId);
+   };
+
+   function recalculateCurrentStageId(entityStageObject) {
+      entityStageObject.currentStageId = currentStageOf(entityStageObject);
    }
 
+   vm.candidateCount = (extStage) => {
+      return extStage.candidates.length ? extStage.candidates.length : '';
+   };
+
+   vm.queryByStage = (extendedStage) => {
+      if (extendedStage) {
+         vm.vacancyStageInfoComposedObjectsToShow = extendedStage.candidates;
+      } else {
+         vm.vacancyStageInfoComposedObjectsToShow = vm.vacancyStageInfosComposedByCandidateIdVacancyId;
+      }
+   };
+
+   vm.callRejectButton = (candidateStage) => {
+      candidateStage.rejectButtonClicked = !!!candidateStage.rejectButtonClicked;
+   };
+
+   vm.goCandidate = (candidateId) => {
+      $state.go('candidateProfile', { _data: null, candidateId });
+   };
+
+   vm.goVacancy = (vacancyId) => {
+      $state.go('vacancyView', { _data: null, vacancyId });
+   };
+
+   vm.filterByStage = (selectedStage) => {
+      selectedStage._isPressed = !selectedStage._isPressed;
+      if (selectedStage._isPressed) {
+         addToQuery(selectedStage);
+      } else {
+         removeFromQuery(selectedStage);
+      }
+      vm.vacancyStageInfoComposedObjectsToShow = needToPerformQuery() ?
+            performQuery() : vm.vacancyStageInfosComposedByCandidateIdVacancyId;
+   };
 
-   function notifySelectedStageChangingBack() {
-      UserDialogService.notification('Stage passing is cancelled', 'warning');
+   function performQuery() {
+      return filter(vm.vacancyStageInfosComposedByCandidateIdVacancyId, (entityStageObject) => {
+         return filter(vm.stageQueries, (extStage) => {
+            return extStage.stage.id === entityStageObject.currentStageId;
+         }).length;
+      });
    }
 
-   function notifyCurrentStageChangingForward() {
-      UserDialogService.notification('Stage was succesfully changed', 'notification');
+   function needToPerformQuery() {
+      return vm.stageQueries.length;
    }
 
-   function _findStagesAndRecomposeBy(stagesObject) {
-      return findStagesBetween(stagesObject).then((stages) => {
-         return recomposeStagesBetween(stages);
+   function addToQuery(selectedStage) {
+      vm.stageQueries = [...vm.stageQueries, selectedStage];
+   }
+   function removeFromQuery(selectedStage) {
+      vm.stageQueries = filter(vm.stageQueries, (extStage) => {
+         return extStage.stage.id !== selectedStage.stage.id;
       });
```

Comment:
it can be next:
 vm.stageQueries = filter(vm.stageQueries, extStage => extStage.stage.id !== selectedStage.stage.id);